untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+$fst = \lambda p. p\;tru,$\
+$snd = \lambda p. p\;fls.$
+
+It's reasonably straightforward to show that these definitions work as intended.
+Consider, for instance,
+
+$fst\;(pair\;x\;y)$
+$= (\lambda p. p\;tru)\;((\lambda f\;s\;b. b\;f\;s)\;x\;y)$
+$\Rightarrow (\lambda p. p\;tru)\;(\lambda b. b\;x\;y)$
+$\to (\lambda b. b\;x\;y)\;tru$
+$\to tru\;x\;y$
+$= (\lambda a\; b. a)\;x\;y$
+$\Rightarrow x.$
+
+Here, double arrow $\Rightarrow$ signifies more than one step of beta-reduction.
+
+### Church naturals
+
+Since we encoded both pairs and booleans, we already can encode naturals simply as sequences of booleans, that is, as binary numbers.
+In practice, this turns out to be cumbersome, however, so instead, we'll define naturals similar to the Peano construction.
+Essentially, we'll be choosing an encoding for zero and then encoding other numbers as some successor function applied to zero.
+
+Church used the following definitions:
+
+$c_0 = \lambda s\;z. z\;$\
+$c_1 = \lambda s\;z. s\;z,$\
+$c_2 = \lambda s\;z. s\;(s\;z),$\
+etc.
+
+The beautiful thing about this encoding is that it naturally encodes the concept of "counting".
+Each Church natural takes a "next" combinator $s$ and a "starting value" $z,$ and then "counts" the corresponding "nexts" from $z.$
+
+Now we can define a successor function, taking a natural and returning the next natural.
+Since we defined naturals as functions of two arguments, we expect the successor function to be a function of three arguments (i.e. three nested abstractions).
+The body should apply the "next" argument to the natural one more time:
+
+$succ = \lambda n\; s\; z. s\;(n\;s\;z).$
+
+It might be easier to see what's going on if we explicitly note the outer abstraction:
+
+$succ = \lambda n. \lambda s\; z. s\;(n\;s\;z).$
+
+In the same manner, we can define addition
+$$plus = \lambda m\; n. \lambda s\; z. m\;s\;(n\;s\;z),$$
+and multiplication
+$$mul = \lambda m\; n. \lambda s\; z. m\;(n\;s)\;z.$$
+The former counts $m$ starting from $n$, and the latter counts $m$ times $n$ starting from $z.$
+
+We can also easily test whether a value corresponds to zero (i.e. $c_0$) or not:
+$$isZero = \lambda n. n\;(\lambda x. fls)\;tru.$$
+
+We run into a bit of an issue when trying to define subtraction or a predecessor function.
+Turns out it's doable but neither pretty nor efficient.
+We'll use pairs of values containing two subsequent naturals.
+That way, we can stagger the counter by one.
+To simplify the notation, we'll introduce two named helper abstractions:
+
+$zz = pair\;c_0\;c_0$\
+$ss = \lambda nn. pair\;(snd\; nn)\;(succ\;(snd\;nn)).$\
+
+The first element of the pair here is the predecessor, and the second is the successor.
+$zz$ encodes the starting point, and $ss$ is the staggered successor.
+
+Now we can define the predecessor function:
+
+$$pred = \lambda n. fst\; (n\; ss\; zz).$$
+
+With this definition, $pred\;c_0 = c_0,$ which isn't ideal.
```

Comment:
I would put the comma outside the latex, It renders two different commas, which isn't very aesthetic, I think.

Reply:
The reason I put commas and other punctuation marks inside LaTeX is that otherwise those punctuation marks will eventually run away to the next line. Arguably, this is worse than a slight font mismatch:

![image](https://user-images.githubusercontent.com/7275622/165609458-40690c60-dc4d-45db-a9b1-73ea531b2a3a.png)

Reply:
It is indeed worse. Okay. 😁